### PR TITLE
perf(sink): don't compact chunk for blackhole sink

### PR DESCRIPTION
**Pull request by xxchan**
I hereby agree to the terms of the [RisingWave Labs, Inc. Contributor License Agreement](https://gist.github.com/TennyZhuang/f00be7f16996ea48effb049aa7be4d66#file-rw_cla).

## What's changed and what's your intention?

motivation: In flamegraph of benchmark q0 (e.g., https://buildkite.com/risingwave-test/nexmark-benchmark/builds/3919#019002e2-0029-49f5-9062-3ea34bac3cb8), there are 3.75% CPU on compact chunk in sink, which seems unnecessary for blackhole sink.

image

## Checklist

- [ ] I have written necessary rustdoc comments
- [ ] I have added necessary unit tests and integration tests
- [ ] I have added test labels as necessary. See [details](https://github.com/risingwavelabs/risingwave/blob/main/docs/developer-guide.md#ci-labels-guide).
- [ ] I have added fuzzing tests or opened an issue to track them. (Optional, recommended for new SQL features #7934).
- [ ] My PR contains breaking changes. (If it deprecates some features, please create a tracking issue to remove them in the future).
- [ ] All checks passed in `./risedev check` (or alias, `./risedev c`)
- [ ] My PR changes performance-critical code. (Please run macro/micro-benchmarks and show the results.)

- [ ] My PR contains critical fixes that are necessary to be merged into the latest release. (Please check out the [details](https://github.com/risingwavelabs/risingwave/blob/main/CONTRIBUTING.md))

## Documentation

- [ ] My PR needs documentation updates. (Please use the **Release note** section below to summarize the impact on users)

## Release note

If this PR includes changes that directly affect users or other significant modifications relevant to the community, kindly draft a release note to provide a concise summary of these changes. Please prioritize highlighting the impact these changes will have on users.

</details>

### CODE REVIEW

File: src/stream/src/executor/sink.rs [hunk 1]
```diff
@@ -184,6 +184,7 @@ impl<F: LogStoreFactory> SinkExecutor<F> {
         let re_construct_with_sink_pk = need_advance_delete
             && self.sink_param.sink_type == SinkType::Upsert
             && !self.sink_param.downstream_pk.is_empty();
+        let is_blackhole = self.sink_param.connector() == Some("blackwhole");
```
Comment (BugenZhao) [inline src/stream/src/executor/sink.rs:187]:
```suggestion
        let is_blackhole = self.sink_param.connector() == Some("blackhole");
```
?

Comment (BugenZhao) [reply src/stream/src/executor/sink.rs:187]: Can we use match the variant of `self.sink` instead (similar to `is_sink_into_table()`)? The current way looks fragile.

File: src/stream/src/executor/sink.rs [hunk 2]
```diff
@@ -300,6 +302,7 @@ impl<F: LogStoreFactory> SinkExecutor<F> {
         input_data_types: Vec<DataType>,
         down_stream_pk: Vec<usize>,
         sink_chunk_buffer_size_metrics: LabelGuardedIntGauge<3>,
+        is_blackhole: bool,
```
Comment (BugenZhao) [inline src/stream/src/executor/sink.rs:305]: Can we make the name more specific, like `compact_chunk`? `is_blackhole` does not look intuitive to me.

**Comment by BugenZhao** [comment]
Please add some context in the PR body.

**Review by st1page — COMMENTED** [review]
Doesn't make a lot of sense to me, blackhole sink is only used in benchmark at the moment, and as far as I know we haven't found any other use for it either. This PR looks a bit like it's being optimized for benchmark. 🤔

**Comment by lmatz** [comment]
> This PR looks a bit like it's being optimized for benchmark.

yes, it is the objective, to publish a few reports for marketing

**Review by st1page — APPROVED** [review]

**Comment by xxchan** [comment]
@BugenZhao description and code updated

**Comment by kwannoel** [comment]
https://github.com/risingwavelabs/risingwave/issues/8712#issuecomment-1482452965

I think this can be applied to all sinks?

**Comment by st1page** [comment]
> [#8712 (comment)](https://github.com/risingwavelabs/risingwave/issues/8712#issuecomment-1482452965)
> 
> I think this can be applied to all sinks? I.e. avoid compact before sink, since we can prune the records when building the encoded value to be sinked.

These two "compactions" have different meanings. What is being discussed here is that within the same chunk, for the same stream key, only all adjacent Insert, Delete pairs are removed, which is a very lightweight operation that only requires changing the visibility bitmap of the chunk. In contrast, the discussion in q0 performance is about reconstructing the chunk to remove all rows with visibility set to false.

**Pull request merged**